### PR TITLE
ENH: concat of nullable int + bool preserves int dtype

### LINKED ISSUE #34095: ENH: enable concat for nullable Int with other dtypes
Following-up on https://github.com/pandas-dev/pandas/pull/33607 which added the `ExtensionDtype._get_common_dtype` method to the EA interface to determine concat/append behaviour, we can now use this to refine the concat behaviour for nullable integer and boolean dtypes.

See comment here: https://github.com/pandas-dev/pandas/pull/33607#discussion_r410127707

For now, the PR only enabled concat of all IntegerDtype objects (any other combination will result in object dtype). But, we can expand this to the combination of IntegerDtype with numpy int dtype, Int with boolean, Int with float, etc

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #34095

### CODE REVIEW

File: pandas/core/arrays/integer.py [hunk 1]
```diff
@@ -92,10 +92,13 @@ def construct_array_type(cls) -> Type["IntegerArray"]:
         return IntegerArray
 
     def _get_common_dtype(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
-        # for now only handle other integer types
+        # we only handle nullable EA dtypes and numeric numpy dtypes
         if not all(
-            isinstance(t, _IntegerDtype)
-            or (isinstance(t, np.dtype) and np.issubdtype(t, np.integer))
+            isinstance(t, BaseMaskedDtype)
```
Comment (TomAugspurger) [inline pandas/core/arrays/integer.py:97]: I worry a bit about this. I could easily see other arrays using the BaseMasked stuff (e.g. StringArray) that shouldn't necessarily be considered "integer-like" for this concat.

So I'd be more comfortable with `(_IntegerDtype, BooleanDtype)` even though those are synonymous today.

**Comment by TomAugspurger** [comment]
Thanks. Does the original whatsnew from the other PR describe this change as well?

File: pandas/core/arrays/integer.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/integer.py:97]: agree with @TomAugspurger here

Comment (jorisvandenbossche) [reply pandas/core/arrays/integer.py:97]: The check for BaseMaskedDtype ensures the dtype has a `umpy_dtype`, and below we still do a `np.find_common_type` on the result. So assuming you have int + string, numpy will return object dtype for that, in which case we still return None from this function.

So even when we make StringArray a masked array, this method should already work as expected.

**Comment by jorisvandenbossche** [comment]
> Thanks. Does the original whatsnew from the other PR describe this change as well?

I think so yes, it's still in general about preserving EA dtype in concat. But I added the issue number to the exsting whatsnew.

File: pandas/core/arrays/integer.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/arrays/integer.py:97]: > So even when we make StringArray a masked array, this method should already work as expected.

Good to know.

>  I don't have to add FloatingDtype to the list of (_IntegerDtype, BooleanDtype) in the floating PR.

Will we want to return None here for float? Or I suppose the `find_common_type` stuff will handle that as well?

Comment (TomAugspurger) [reply pandas/core/arrays/integer.py:97]: @jreback does https://github.com/pandas-dev/pandas/pull/34985#discussion_r445670070 make sense?

**Review by TomAugspurger — APPROVED** [review]

File: pandas/core/arrays/integer.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/integer.py:97]: make sense, though still feel that there is no downside to being explict about listing the 2 dtypes directy; its more obvious. Agreed that if in the future we add more convertable to integer dtypes they won't automatically be added, but i think that is of lesser benefit that better readability here (i mean you could add a comment, but i think listing the classes is better)

Comment (jorisvandenbossche) [reply pandas/core/arrays/integer.py:97]: I think how it is written now also gives a good message: all BaseMaskedDtype subclasses are supported by this method (which *is* the case, even though some might return `None` from the function later on)

> Will we want to return None here for float? Or I suppose the find_common_type stuff will handle that as well, just like string?

Yes, right now we would still return None for float (only if the common numpy dtype is an integer dtype, an EA dtype is returned). Once we have FloatingArray, we would add an additional check for the case that the common numpy dtype is a float dtype, and then return an EA floating dtype.

**Comment by TomAugspurger** [comment]
Merged master.

@jreback was the discussion in https://github.com/pandas-dev/pandas/pull/34985#discussion_r449643794 resolved?

**Comment by jreback** [comment]
yes this is fine to merge

**Pull request merged**